File: Linked Lists/Merge Two Sorted Lists/MergeTwoSortedLists.java
/**
    Definition for singly-linked list.
    public class ListNode {
        int val;
        ListNode next;
        ListNode() {

        }

        ListNode(int val) { 
            this.val = val; 
        }

        ListNode(int val, ListNode next) { 
            this.val = val; this.next = next; 
        }
    }
 */


 class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        
        ListNode merged = new ListNode();

        if(list1 == null){
            merged = list2;
        }
        else if(list2 == null){
            merged = list1;
        }
        else if(list1.val <= list2.val){
            merged = list1;
            merged.next = mergeTwoLists(merged.next, list2);  
        }
        else if(list1.val > list2.val){
            merged = list2;
            merged.next = mergeTwoLists(list1, merged.next);  
        }

        return merged; 

        //remove the lowest and add to the merged
        //call the function again with both list Node next node
    }
}

// 1 

File: Linked Lists/Merge Two Sorted Lists/mergeTwoSortedLists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


# iterative approach
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1 and not list2:
            return list1
        elif not list1:
            return list2
        elif not list2:
            return list1

        curr1 = list1
        curr2 = list2
        if curr1.val <= curr2.val:
            sorted = curr1
            resultHead = sorted
            curr1 = curr1.next
        else:
            sorted = curr2
            resultHead = sorted
            curr2 = curr2.next

        while(curr1 is not None and curr2 is not None):
            if curr1.val <= curr2.val:
                sorted.next = curr1
                sorted = sorted.next
                curr1 = curr1.next
            else:
                sorted.next = curr2
                sorted = sorted.next
                curr2 = curr2.next

        if curr1 is None:
            sorted.next = curr2
        elif curr2 is None:
            sorted.next = curr1

        return resultHead
    

#recursive approach